fdbserver/RestoreLoader.actor.cpp

```diff
@@ -518,27 +531,39 @@ ACTOR static Future<Void> _parseRangeFileToMutationsOnLoader(VersionedMutationsM
 // pair.second of pairs with the same pair.first.
-ACTOR static Future<Void> _parseLogFileToMutationsOnLoader(
-    std::map<Standalone<StringRef>, Standalone<StringRef>>* pMutationMap,
-    std::map<Standalone<StringRef>, uint32_t>* pMutationPartMap, Reference<IBackupContainer> bc, Version version,
-    std::string fileName, int64_t readOffset, int64_t readLen, KeyRange restoreRange, Key addPrefix, Key removePrefix,
-    Key mutationLogPrefix) {
+ACTOR static Future<Void> _parseLogFileToMutationsOnLoader(NotifiedVersion* pProcessedFileOffset,
+                                                           SerializedMutationListMap* pMutationMap,
+                                                           SerializedMutationPartMap* pMutationPartMap,
+                                                           Reference<IBackupContainer> bc, Version version,
+                                                           std::string fileName, int64_t readOffset, int64_t readLen,
+                                                           KeyRange restoreRange, Key addPrefix, Key removePrefix,
+                                                           Key mutationLogPrefix) {
 	state Reference<IAsyncFile> inFile = wait(bc->readFile(fileName));
 	// decodeLogFileBlock() must read block by block!
 	state Standalone<VectorRef<KeyValueRef>> data =
 	    wait(parallelFileRestore::decodeLogFileBlock(inFile, readOffset, readLen));
-	TraceEvent("FastRestore").detail("DecodedLogFile", fileName).detail("DataSize", data.contents().size());
-
-	state int start = 0;
-	state int end = data.size();
-	state int numConcatenated = 0;
-	for (int i = start; i < end; ++i) {
-		//Key k = data[i].key.withPrefix(mutationLogPrefix);
-		//ValueRef v = data[i].value;
-		// Concatenate the backuped param1 and param2 (KV) at the same version.
-		bool concatenated =
-		    concatenateBackupMutationForLogFile(pMutationMap, pMutationPartMap, data[i].key, data[i].value);
-		numConcatenated += (concatenated ? 1 : 0);
+	TraceEvent("FastRestore")
+	    .detail("DecodedLogFile", fileName)
+	    .detail("Offset", readOffset)
+	    .detail("Length", readLen)
+	    .detail("DataSize", data.contents().size());
+
+	// Ensure data blocks in the same file are processed in order
+	wait(pProcessedFileOffset->whenAtLeast(readOffset));
+
+	if (pProcessedFileOffset->get() == readOffset) {
+		state int start = 0;
+		state int end = data.size();
+		state int numConcatenated = 0;
+		for (int i = start; i < end; ++i) {
+			// Key k = data[i].key.withPrefix(mutationLogPrefix);
+			// ValueRef v = data[i].value;
+			// Concatenate the backuped param1 and param2 (KV) at the same version.
+			bool concatenated =
+			    concatenateBackupMutationForLogFile(pMutationMap, pMutationPartMap, data[i].key, data[i].value);
+			numConcatenated += (concatenated ? 1 : 0);
+		}
+		pProcessedFileOffset->set(readOffset + readLen);
```

Comment:
This is pretty clearly using a notified version for ... not the version.  I'm okay with merging this, but I'd appreciate if you can look into how much work it would be to refactor NotifiedVersion into a Notified<Version> so that you could just use a Notified<long> here instead.

Reply:
Is there a Notified struct? 
I didn't find `struct Notified` in the code base.
Did you suggest adding a new Notified type or
abstract the Notified concept and use it to implement the actual notified type, say `NotifiedVersion` and `NotifiedDouble`?